### PR TITLE
[WIP] Ortmodule onnx input support

### PR DESCRIPTION
**Pull request by almaly-microsoft**
ORTmodule wraps pytorch model and executes training for it. However pytorch models are stored in ONNX format and in some cases (like quantization emulation) they needs to be retrained to get accuracy back.

This change is about adding that support. I updated one test to store pytorch model into ONNX format and then passed it into ORTModule. Training runs well.

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 1]
```diff
@@ -88,29 +96,33 @@ def __init__(self, module):
...
+                                      UserWarning)
+        else:
+            self._module_parameters = [InputParameters(i.name, inspect.Parameter.POSITIONAL_OR_KEYWORD) \
+                    for i in self._onnx_model.graph.input]
+        
 
         self.is_rocm_pytorch = (True if ((torch.version.hip is not None) and (ROCM_HOME is not None)) else False)
 
-        self._use_external_gpu_allocator = True
+        self._use_external_gpu_allocator = False
```
Comment (almaly-microsoft) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:125]: Rollback

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 2]
```diff
@@ -88,29 +96,33 @@ def __init__(self, module):
         self._use_static_shape = False
 
         # flag to enable symbolic shape inference for dynamic shape inputs to improve performance
-        self._run_symbolic_shape_infer = True
+        self._run_symbolic_shape_infer = False
```
Comment (almaly-microsoft) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:99]: Rollback

File: orttraining/orttraining/python/training/ortmodule/_io.py [hunk 3]
```diff
@@ -69,8 +67,7 @@ def unflatten(self, flat_args):
         '''Unflatten tuple of tensors into args and kwargs'''
 
         args = tuple(flat_args[:self.num_positionals])
-        kwargs = {name: arg for name, arg in zip(self.names[self.num_positionals_non_none:], flat_args[self.num_positionals:]) \
-            if name in self.keyword_names}
+        kwargs = {kwarg_name: arg for kwarg_name, arg in zip(self.keyword_names, flat_args[self.num_positionals:])}
```
Comment (almaly-microsoft) [inline orttraining/orttraining/python/training/ortmodule/_io.py:70]: check why that happened. didn't change that

**Review by mrry — CHANGES_REQUESTED** [review]
Sorry for the delay on this review, and thanks for bearing with us while we were doing the release, Aleksandr.

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py [hunk 4]
```diff
@@ -188,10 +190,11 @@ def backward(ctx, *grad_outputs):
                                         _ORTModuleFunction.apply(
                                             *_io._combine_input_buffers_initializers(
                                                 [param for name, param in self._flattened_module.named_parameters()
-                                                    if name in self._graph_info.initializer_names],
+                                                    if name in self._graph_info.initializer_names] if self._flattened_module else \
+                                                    [p[1] for p in self._onnx_model_parameters],
```
[the remark this answers is not in the document]
Comment (almaly-microsoft) [reply orttraining/orttraining/python/training/ortmodule/_training_manager.py:194]: merged in, thanks

File: orttraining/orttraining/python/training/ortmodule/ortmodule.py [hunk 5]
```diff
@@ -40,46 +42,67 @@ class ORTModule(torch.nn.Module):
...
+                  model's output.
+               - a combined ONNX model and loss function.
+            device: device to store tensors (e.g. 'cpu', 'cuda', 'cuda:<int_idx>').
+        """
 
-        # Bind the forward method.
-        self.forward = _forward.__get__(self)
-        # Copy the forward signature from the PyTorch module.
-        functools.update_wrapper(
-            self.forward.__func__, module.forward.__func__)
```
[the remark this answers is not in the document]
Comment (almaly-microsoft) [reply orttraining/orttraining/python/training/ortmodule/ortmodule.py:47]: it is intentional as onnx graph doesn't have forward method

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 6]
```diff
@@ -287,11 +307,16 @@ def _initialize_graph_builder(self, training):
...
+        initializer_names = []
+        initializer_names_to_train = []
+        if self._flattened_module:
+            # TODO: PyTorch exporter bug: changes the initializer order in ONNX model
+            initializer_names = [name for name, _ in self._flattened_module.named_parameters()]
+            initializer_names_to_train = [name for name, param in self._flattened_module.named_parameters() 
+                                          if param.requires_grad and name in onnx_initializer_names]
+        elif self._onnx_model:
+            initializer_names = [initializer.name for initializer in self._onnx_model.graph.initializer]
+            initializer_names_to_train = [initializer.name for initializer in self._onnx_model.graph.initializer]
```
[the remark this answers is not in the document]
Comment (almaly-microsoft) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:340]: I haven't tried anything that is more advanced that model in orttraining_test_ortmodule_poc.py

**Comment by thiagocrepaldi** [comment]
I will start analyzing this PR soon and propose changes to both ORTModule and this PR so that this PR can be plugged in a modular way

**Comment by thiagocrepaldi** [comment]
closed due to inactivity
